File: src/main/java/com/babjo/whatdaybot/crawler/RisingKeywordCrawler.java
package com.babjo.whatdaybot.crawler;

import static com.google.common.collect.ImmutableList.toImmutableList;

import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.babjo.whatdaybot.model.RisingKeyword;
import com.babjo.whatdaybot.utils.URIUtils;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class RisingKeywordCrawler {

    private static final Logger logger = LoggerFactory.getLogger(RisingKeywordCrawler.class);

    public List<RisingKeyword> doCrawling(LocalDateTime dateTime) throws IOException {
        Document doc = Jsoup.connect("https://datalab.naver.com/keyword/realtimeList.naver?datetime=" + dateTime
                .format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"))).get();
        logger.info(doc.outerHtml());

        Element element = doc.select(".select_date").get(0);
        Elements keywords = element.select(".title");

        List<RisingKeyword> result = keywords.stream()
                                             .map(Element::text)
                                             .map(text -> new RisingKeyword(text, String.format(
                                                     "https://search.naver.com/search.naver?query=%s",
                                                     URIUtils.encodeURIComponent(text))))
                                             .collect(toImmutableList());

        logger.info("doCrawling results: {}", result);

        return result;
    }
}

File: src/main/kotlin/com/babjo/whatdaybot/crawler/RisingKeywordCrawler.kt
package com.babjo.whatdaybot.crawler

import com.babjo.whatdaybot.RisingKeyword
import com.babjo.whatdaybot.naver.openapi.URLShortenerClient
import io.reactivex.Single
import mu.KotlinLogging
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.time.Clock
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeUnit

class RisingKeywordCrawler(
    private val clock: Clock,
    private val timeoutSeconds: Long,
    private val urlShortenerClient: URLShortenerClient
) {

    private val logger = KotlinLogging.logger {}
    lateinit var latestRefreshTime: LocalDateTime
    lateinit var latestRisingKeywords: List<RisingKeyword>

    fun refresh() {
        val target = getTargetTime(LocalDateTime.now(clock))

        doCrawling(target)
            .flatMap { Single.merge(it.map(::shortenUrl)).toList() }
            .timeout(timeoutSeconds, TimeUnit.SECONDS)
            .subscribe { keywords, t ->
                if (t != null) {
                    logger.error("Failed to refresh", t)
                } else {
                    latestRisingKeywords = keywords
                    latestRefreshTime = target
                }
            }
    }

    private fun getTargetTime(now: LocalDateTime) =
        now.let {
            var sec = it.second
            sec %= 30
            it.plusSeconds((-sec).toLong())
        }

    private fun shortenUrl(keyword: RisingKeyword): Single<RisingKeyword> =
        urlShortenerClient
            .shorten(keyword.url)
            .map { RisingKeyword(keyword.text, it.result.url) }
            .onErrorReturn { keyword }

    private fun doCrawling(dateTime: LocalDateTime): Single<List<RisingKeyword>> =
        Single.fromCallable {
            Jsoup
                .connect(
                    "https://datalab.naver.com/keyword/realtimeList.naver?datetime=" +
                            dateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"))
                )
                .get()
                .select(".select_date")[0]
                .select(".title")
                .map(Element::text)
                .map {
                    RisingKeyword(
                        it,
                        "https://search.naver.com/search.naver?query=${encodeURIComponent(it)}"
                    )
                }
        }

    private fun encodeURIComponent(s: String) = try {
        URLEncoder.encode(s, "UTF-8")
            .replace("\\+".toRegex(), "%20")
            .replace("\\%21".toRegex(), "!")
            .replace("\\%27".toRegex(), "'")
            .replace("\\%28".toRegex(), "(")
            .replace("\\%29".toRegex(), ")")
            .replace("\\%7E".toRegex(), "~")
    } catch (e: UnsupportedEncodingException) {
        s
    }
}
